doppelganger/doppelganger.go

```diff
@@ -0,0 +1,325 @@
+		}
+	}
+
+	// Remove validators that are no longer in the current set
+	for idx := range h.validatorsState {
+		if _, exists := retrievedValidatorsSet[idx]; !exists {
+			removedValidators = append(removedValidators, uint64(idx))
+			delete(h.validatorsState, idx)
+		}
+	}
+
+	if len(addedValidators) > 0 {
+		h.logger.Debug("Added validators to Doppelganger state", zap.Uint64s("validator_indices", addedValidators))
+	}
+
+	if len(removedValidators) > 0 {
+		h.logger.Debug("Removed validators from Doppelganger state", zap.Uint64s("validator_indices", removedValidators))
+	}
+}
+
+// RemoveValidatorState removes the validator from the Doppelganger state.
+func (h *handler) RemoveValidatorState(validatorIndex phase0.ValidatorIndex) {
+	h.mu.Lock()
+	defer h.mu.Unlock()
+
+	if _, exists := h.validatorsState[validatorIndex]; !exists {
+		h.logger.Warn("Validator not found in Doppelganger state", fields.ValidatorIndex(validatorIndex))
+		return
+	}
+
+	delete(h.validatorsState, validatorIndex)
+	h.logger.Debug("Removed validator from Doppelganger state", fields.ValidatorIndex(validatorIndex))
+}
+
+// Start starts the Doppelganger monitoring.
+func (h *handler) Start(ctx context.Context) error {
+	h.logger.Info("Doppelganger monitoring started")
+
+	var startEpoch, previousEpoch phase0.Epoch
+	firstRun := true
+	ticker := h.slotTickerProvider()
+	slotsPerEpoch := h.network.Beacon.SlotsPerEpoch()
+
+	for {
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		case <-ticker.Next():
+			currentSlot := ticker.Slot()
+			currentEpoch := h.network.Beacon.EstimatedEpochAtSlot(currentSlot)
+
+			buildStr := fmt.Sprintf("e%v-s%v-#%v", currentEpoch, currentSlot, currentSlot%32+1)
+			h.logger.Debug("🛠 ticker event", zap.String("epoch_slot_pos", buildStr))
+
+			// Update DG state with self participating validators from validator provider at the current epoch
+			validatorIndices := indicesFromShares(h.validatorProvider.SelfParticipatingValidators(currentEpoch))
+			h.updateDoppelgangerState(validatorIndices)
+
+			// Perform liveness checks during the first run or at the last slot of the epoch.
+			// This ensures that the beacon node has had enough time to observe blocks and attestations,
+			// preventing delays in marking a validator as safe.
+			if (!firstRun && uint64(currentSlot)%slotsPerEpoch != slotsPerEpoch-1) || startEpoch == currentEpoch {
+				continue
+			}
+
+			if firstRun {
+				startEpoch = currentEpoch
+				firstRun = false
+			}
+
+			// Detect if an unexpected gap in epochs occurred (e.g., due to system sleep or clock drift).
+			// If we detect a skipped epoch, we reset all doppelganger states to avoid unsafe signing.
+			if previousEpoch > 0 && currentEpoch != previousEpoch+1 {
+				h.logger.Warn("Epoch skipped unexpectedly, resetting all Doppelganger states",
+					zap.Uint64("previous_epoch", uint64(previousEpoch)),
+					zap.Uint64("current_epoch", uint64(currentEpoch)),
+				)
+
+				h.resetDoppelgangerStates()
```

Comment:
I'll add a small comment that will say this means our operator will probably skip at least a few slots (assuming it sees quorum next slot or something)